File: data/src/main/java/org/cryptomator/data/cloud/s3/S3Folder.java
package org.cryptomator.data.cloud.s3;

import org.cryptomator.domain.Cloud;
import org.cryptomator.domain.CloudFolder;

class S3Folder implements CloudFolder, S3Node {

	private static final String DELIMITER = "/";

	private final S3Folder parent;
	private final String name;
	private final String path;

	public S3Folder(S3Folder parent, String name, String path) {
		this.parent = parent;
		this.name = name;
		this.path = path;
	}

	@Override
	public Cloud getCloud() {
		return parent.getCloud();
	}

	@Override
	public String getName() {
		return name;
	}

	@Override
	public String getPath() {
		return path;
	}

	@Override
	public String getKey() {
		if (path.startsWith(DELIMITER)) {
			return path.substring(DELIMITER.length()) + DELIMITER;
		}
		return path + DELIMITER;
	}

	@Override
	public S3Folder getParent() {
		return parent;
	}

	@Override
	public S3Folder withCloud(Cloud cloud) {
		return new S3Folder(parent.withCloud(cloud), name, path);
	}
}

File: data/src/main/java/org/cryptomator/data/cloud/s3/S3Folder.kt
package org.cryptomator.data.cloud.s3

import org.cryptomator.domain.Cloud
import org.cryptomator.domain.CloudFolder

internal open class S3Folder(override val parent: S3Folder?, override val name: String, override val path: String) : CloudFolder, S3Node {

	override val cloud: Cloud?
		get() = parent?.cloud

	override val key: String
		get() = if (path.startsWith(DELIMITER)) {
			path.substring(DELIMITER.length) + DELIMITER
		} else path + DELIMITER

	override fun withCloud(cloud: Cloud?): S3Folder? {
		return S3Folder(parent?.withCloud(cloud), name, path)
	}

	companion object {
		private const val DELIMITER = "/"
	}
}
